### ISSUE TITLE
blip_captioning returns "is_url" is not defined

### ISSUE THREAD

**Issue by Memphiiisto** [bmaltais/kohya_ss#1820]
Hi there!
I tried to Blip Caption some of my images.
Sadly it returns an error:
NameError: name 'is_url' is not defined

`16:55:30-908782 INFO     Captioning files in C:/TrainingImages...
16:55:30-909781 INFO     ./venv/Scripts/python.exe "finetune/make_captions.py" --batch_size="1" --num_beams="1"
                         --top_p="0.9" --max_length="75" --min_length="5" --beam_search --caption_extension=".txt"
                         "C:/TrainingImages"
                         --caption_weights="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/mode
                         l_large_caption.pth"
Current Working Directory is:  C:\Users\Marcel\Desktop\kohya_ss
Traceback (most recent call last):
  File "C:\Users\{USER}\Desktop\kohya_ss\finetune\make_captions.py", line 202, in <module>
    main(args)
  File "C:\Users\{USER}\Desktop\kohya_ss\finetune\make_captions.py", line 79, in main
    if not is_url(args.caption_weights) and not os.path.isfile(args.caption_weights):
NameError: name 'is_url' is not defined
16:55:35-457706 INFO     ...captioning done`

anyone got an idea what i can do?
Thanks!

i just installed it new. same problem

**Comment by Memphiiisto** [comment]
I was able to get it to work myself.
I copied the is_url() function from a different file (blip.py) into the "make_captions.py".

first was an import:
`from urllib.parse import urlparse`

and then the func:
```
def is_url(url_or_filename):
    parsed = urlparse(url_or_filename)
    return parsed.scheme in ("http", "https")
```

hope that helps if anyone got the same problem.

Thanks
Mem

**Comment by cain666-4u** [comment]
i have the same issue, but im not sure where the files are you changed.. i cant seem to find them

**Comment by paboum** [comment]
> I was able to get it to work myself. I copied the is_url() function from a different file (blip.py) into the "make_captions.py".

It seems unnecessary to copy that code while you can just import it. That's what I did in https://github.com/bmaltais/kohya_ss/pull/1821 - hopefully you can confirm if that works for you too.

**Comment by boricuapab** [comment]
> > I was able to get it to work myself. I copied the is_url() function from a different file (blip.py) into the "make_captions.py".
> 
> It seems unnecessary to copy that code while you can just import it. That's what I did in #1821 - hopefully you can confirm if that works for you too.

Thanks, this worked

error

![sdxlBlipError](https://github.com/bmaltais/kohya_ss/assets/61924816/18cb3127-6159-4533-93d8-7043c9684be0)

fixed

![sdxlBlipErrorFixed](https://github.com/bmaltais/kohya_ss/assets/61924816/284069e3-4c3e-4b49-b220-e5ccf9201d37)

**Comment by cain666-4u** [comment]
both the url check and the blip import look present in the make_captions.py file on my system, but its not working? what have i missed? being a noob on python and code in general im a bit lost on what i need to fix?

**Comment by stasha17** [comment]
Hi.  I found the .py file, but where do you "type" that code? Both the import and the func?

**Comment by boricuapab** [comment]
![addThis](https://github.com/bmaltais/kohya_ss/assets/61924816/3b50b2fa-c3b5-4985-b985-69997b215ae9)

add what I underlined in red in the image above to your make_captions.py and you're good

**Comment by metalons** [comment]
23:39:33-541591 INFO     Start training LoRA Standard ...
23:39:33-543589 INFO     Checking for duplicate image filenames in training data directory...
23:39:33-545589 INFO     Valid image folder names found in: C:/Users/metal/Desktop/LoraTest 1920x1080/image
23:39:33-547615 INFO     Folder 150_Make: 15 images found
23:39:33-548591 INFO     Folder 150_Make: 2250 steps
23:39:33-549591 INFO     Total steps: 2250
23:39:33-550590 INFO     Train batch size: 1
23:39:33-551590 INFO     Gradient accumulation steps: 1
23:39:33-552650 INFO     Epoch: 1
23:39:33-553673 INFO     Regulatization factor: 1
23:39:33-554675 INFO     max_train_steps (2250 / 1 / 1 * 1 * 1) = 2250
23:39:33-555652 INFO     stop_text_encoder_training = 0
23:39:33-556726 INFO     lr_warmup_steps = 225
23:39:33-557755 INFO     Saving training config to C:/Users/metal/Desktop/LoraTest
                         1920x1080/model\make_20231231-233933.json...
23:39:33-558754 INFO     accelerate launch --num_cpu_threads_per_process=2 "./train_network.py" --enable_bucket
                         --min_bucket_reso=256 --max_bucket_reso=2048
                         --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5"
                         --train_data_dir="C:/Users/metal/Desktop/LoraTest 1920x1080/image" --resolution="512,512"
                         --output_dir="C:/Users/metal/Desktop/LoraTest 1920x1080/model"
                         --logging_dir="C:/Users/metal/Desktop/LoraTest 1920x1080/log" --network_alpha="1"
                         --save_model_as=safetensors --network_module=networks.lora --text_encoder_lr=5e-05
                         --unet_lr=0.0001 --network_dim=8 --output_name="make" --lr_scheduler_num_cycles="1"
                         --no_half_vae --learning_rate="0.0001" --lr_scheduler="cosine" --lr_warmup_steps="225"
                         --train_batch_size="1" --max_train_steps="2250" --save_every_n_epochs="1"
                         --mixed_precision="fp16" --save_precision="fp16" --cache_latents --optimizer_type="AdamW8bit"
                         --max_grad_norm="1" --max_data_loader_n_workers="0" --bucket_reso_steps=64 --mem_eff_attn
                         --gradient_checkpointing --xformers --bucket_no_upscale --noise_offset=0.0
Traceback (most recent call last):
  File "C:\Users\metal\kohya_ss\train_network.py", line 30, in <module>
    import library.train_util as train_util
  File "C:\Users\metal\kohya_ss\library\train_util.py", line 22
    from accelerate import Accelerator, InitProcessGroupKwargs, DistributedDataParallelKwargs,
                                                                                               ^
SyntaxError: trailing comma not allowed without surrounding parentheses
Traceback (most recent call last):
  File "C:\Users\metal\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\metal\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\metal\kohya_ss\venv\Scripts\accelerate.exe\__main__.py", line 7, in <module>
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\accelerate_cli.py", line 47, in main
    args.func(args)
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\launch.py", line 986, in launch_command
    simple_launcher(args)
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\launch.py", line 628, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['C:\\Users\\metal\\kohya_ss\\venv\\Scripts\\python.exe', './train_network.py', '--enable_bucket', '--min_bucket_reso=256', '--max_bucket_reso=2048', '--pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5', '--train_data_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/image', '--resolution=512,512', '--output_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/model', '--logging_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/log', '--network_alpha=1', '--save_model_as=safetensors', '--network_module=networks.lora', '--text_encoder_lr=5e-05', '--unet_lr=0.0001', '--network_dim=8', '--output_name=make', '--lr_scheduler_num_cycles=1', '--no_half_vae', '--learning_rate=0.0001', '--lr_scheduler=cosine', '--lr_warmup_steps=225', '--train_batch_size=1', '--max_train_steps=2250', '--save_every_n_epochs=1', '--mixed_precision=fp16', '--save_precision=fp16', '--cache_latents', '--optimizer_type=AdamW8bit', '--max_grad_norm=1', '--max_data_loader_n_workers=0', '--bucket_reso_steps=64', '--mem_eff_attn', '--gradient_checkpointing', '--xformers', '--bucket_no_upscale', '--noise_offset=0.0']' returned non-zero exit status 1.

Hi!, Then, can you tell me, what is the problem in this codes???

**Comment by metalons** [comment]
prepare optimizer, data loader etc.
Traceback (most recent call last):
  File "C:\Users\metal\kohya_ss\library\train_util.py", line 3480, in get_optimizer
    import bitsandbytes as bnb
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\__init__.py", line 6, in <module>
    from . import cuda_setup, utils, research
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\research\__init__.py", line 1, in <module>
    from . import nn
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\research\nn\__init__.py", line 1, in <module>
    from .modules import LinearFP8Mixed, LinearFP8Global
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\research\nn\modules.py", line 8, in <module>
    from bitsandbytes.optim import GlobalOptimManager
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\optim\__init__.py", line 6, in <module>
    from bitsandbytes.cextension import COMPILED_WITH_CUDA
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\cextension.py", line 5, in <module>
    from .cuda_setup.main import evaluate_cuda_setup
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\cuda_setup\main.py", line 21, in <module>
    from .paths import determine_cuda_runtime_lib_path
ModuleNotFoundError: No module named 'bitsandbytes.cuda_setup.paths'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\metal\kohya_ss\train_network.py", line 996, in <module>
    trainer.train(args)
  File "C:\Users\metal\kohya_ss\train_network.py", line 348, in train
    optimizer_name, optimizer_args, optimizer = train_util.get_optimizer(args, trainable_params)
  File "C:\Users\metal\kohya_ss\library\train_util.py", line 3482, in get_optimizer
    raise ImportError("No bitsandbytes / bitsandbytesがインストールされていないようです")
ImportError: No bitsandbytes / bitsandbytesがインストールされていないようです
Traceback (most recent call last):
  File "C:\Users\metal\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\metal\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\metal\kohya_ss\venv\Scripts\accelerate.exe\__main__.py", line 7, in <module>
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\accelerate_cli.py", line 47, in main
    args.func(args)
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\launch.py", line 986, in launch_command
    simple_launcher(args)
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\launch.py", line 628, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['C:\\Users\\metal\\kohya_ss\\venv\\Scripts\\python.exe', './train_network.py', '--enable_bucket', '--min_bucket_reso=256', '--max_bucket_reso=2048', '--pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5', '--train_data_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/image', '--resolution=512,512', '--output_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/model', '--logging_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/log', '--network_alpha=1', '--save_model_as=safetensors', '--network_module=networks.lora', '--text_encoder_lr=5e-05', '--unet_lr=0.0001', '--network_dim=8', '--output_name=Danah', '--lr_scheduler_num_cycles=1', '--no_half_vae', '--learning_rate=0.0001', '--lr_scheduler=cosine', '--lr_warmup_steps=225', '--train_batch_size=1', '--max_train_steps=2250', '--save_every_n_epochs=1', '--mixed_precision=fp16', '--save_precision=fp16', '--cache_latents', '--optimizer_type=AdamW8bit', '--max_grad_norm=1', '--max_data_loader_n_workers=0', '--bucket_reso_steps=64', '--mem_eff_attn', '--gradient_checkpointing', '--xformers', '--bucket_no_upscale', '--noise_offset=0.0']' returned non-zero exit status 1.

Can you explain this problem???? :)

**Comment by metalons** [comment]
prepare optimizer, data loader etc.
Traceback (most recent call last):
  File "C:\Users\metal\kohya_ss\library\train_util.py", line 3480, in get_optimizer
    import bitsandbytes as bnb
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\__init__.py", line 6, in <module>
    from . import cuda_setup, utils, research
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\research\__init__.py", line 1, in <module>
    from . import nn
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\research\nn\__init__.py", line 1, in <module>
    from .modules import LinearFP8Mixed, LinearFP8Global
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\research\nn\modules.py", line 8, in <module>
    from bitsandbytes.optim import GlobalOptimManager
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\optim\__init__.py", line 6, in <module>
    from bitsandbytes.cextension import COMPILED_WITH_CUDA
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\cextension.py", line 5, in <module>
    from .cuda_setup.main import evaluate_cuda_setup
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\bitsandbytes\cuda_setup\main.py", line 21, in <module>
    from .paths import determine_cuda_runtime_lib_path
ModuleNotFoundError: No module named 'bitsandbytes.cuda_setup.paths'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\metal\kohya_ss\train_network.py", line 996, in <module>
    trainer.train(args)
  File "C:\Users\metal\kohya_ss\train_network.py", line 348, in train
    optimizer_name, optimizer_args, optimizer = train_util.get_optimizer(args, trainable_params)
  File "C:\Users\metal\kohya_ss\library\train_util.py", line 3482, in get_optimizer
    raise ImportError("No bitsandbytes / bitsandbytesがインストールされていないようです")
ImportError: No bitsandbytes / bitsandbytesがインストールされていないようです
Traceback (most recent call last):
  File "C:\Users\metal\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\metal\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\metal\kohya_ss\venv\Scripts\accelerate.exe\__main__.py", line 7, in <module>
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\accelerate_cli.py", line 47, in main
    args.func(args)
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\launch.py", line 986, in launch_command
    simple_launcher(args)
  File "C:\Users\metal\kohya_ss\venv\lib\site-packages\accelerate\commands\launch.py", line 628, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['C:\\Users\\metal\\kohya_ss\\venv\\Scripts\\python.exe', './train_network.py', '--enable_bucket', '--min_bucket_reso=256', '--max_bucket_reso=2048', '--pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5', '--train_data_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/image', '--resolution=512,512', '--output_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/model', '--logging_dir=C:/Users/metal/Desktop/LoraTest 1920x1080/log', '--network_alpha=1', '--save_model_as=safetensors', '--network_module=networks.lora', '--text_encoder_lr=5e-05', '--unet_lr=0.0001', '--network_dim=8', '--output_name=test', '--lr_scheduler_num_cycles=1', '--no_half_vae', '--learning_rate=0.0001', '--lr_scheduler=cosine', '--lr_warmup_steps=225', '--train_batch_size=1', '--max_train_steps=2250', '--save_every_n_epochs=1', '--mixed_precision=fp16', '--save_precision=fp16', '--cache_latents', '--optimizer_type=AdamW8bit', '--max_grad_norm=1', '--max_data_loader_n_workers=0', '--bucket_reso_steps=64', '--mem_eff_attn', '--gradient_checkpointing', '--xformers', '--bucket_no_upscale', '--noise_offset=0.0']' returned non-zero exit status 1.


HI! Can you tell me what is this problem???? :)

**Comment by paboum** [comment]
If you guys aren't that much into programming and cannot find/edit the file yourself, maybe it will be easier to just use the version with my patch until it becomes merged/resolved in other way. It should match v22.4.0 and I don't plan to keep it up-to-date. And in general, you shouldn't just download random files from strangers but I'm not forcing you.
https://codeload.github.com/paboum/kohya_ss/zip/refs/heads/master

**Comment by Memphiiisto** [comment]
> s unnecessary to copy that code while you can just import it. That's what

Sorry for the late reply!
Thank you! :-)

I shurely will try out your solution as well!
If i dont have to copy the code and can just import it that sounds great haha!
but that will happen next year... so probably in a few hours ;-)

thanks and have a great new year!
Memphisto

**Comment by metalons** [comment]
-----Original Message-----

**Comment by metalons** [comment]
-----Original Message-----

**Comment by ShadowTheFrog** [comment]
> If you guys aren't that much into programming and cannot find/edit the file yourself, maybe it will be easier to just use the version with my patch until it becomes merged/resolved in other way. It should match v22.4.0 and I don't plan to keep it up-to-date. And in general, you shouldn't just download random files from strangers but I'm not forcing you. https://codeload.github.com/paboum/kohya_ss/zip/refs/heads/master

it worked for me, thank you

**Comment by paboum** [comment]
My fix was merged in v22.4.1 and thus I've removed the fork.
If you still have the issue, please update to the recent version.
I suggest closing this ticket if it works now. @Memphiiisto

**Comment by Memphiiisto** [comment]
Oh sorry! I didnt know i had to close it myself :-)
Will do! The change worked for me! 

Thanks everyone
Mem
